web/src/lib/data-hooks/useUnsavedChangesGuard.ts

```diff
@@ -0,0 +1,111 @@
+import { useRouter } from "next/compat/router";
+import { useCallback, useEffect, useRef, useState } from "react";
+
+interface UseUnsavedChangesGuardProps {
+  hasUnsavedChanges: boolean;
+}
+
+interface UseUnsavedChangesGuardReturn {
+  isBlocked: boolean;
+  pendingUrl: string | null;
+  allowNavigation: () => void;
+  blockNavigation: () => void;
+  reset: () => void;
+}
+
+export const useUnsavedChangesGuard = ({
+  hasUnsavedChanges,
+}: UseUnsavedChangesGuardProps): UseUnsavedChangesGuardReturn => {
+  const router = useRouter();
+  const [isBlocked, setIsBlocked] = useState(false);
+  const [pendingUrl, setPendingUrl] = useState<string | null>(null);
+  const isNavigationAllowedRef = useRef(false);
+
+  const hasUnsavedChangesRef = useRef(hasUnsavedChanges);
+  hasUnsavedChangesRef.current = hasUnsavedChanges;
+
+  const allowNavigation = useCallback(() => {
+    isNavigationAllowedRef.current = true;
+    setIsBlocked(false);
+    if (pendingUrl && router) {
+      router.push(pendingUrl);
+    }
+    setPendingUrl(null);
+  }, [pendingUrl, router]);
+
+  const blockNavigation = useCallback(() => {
+    setIsBlocked(false);
+    setPendingUrl(null);
+    isNavigationAllowedRef.current = false;
+  }, []);
+
+  const reset = useCallback(() => {
+    setIsBlocked(false);
+    setPendingUrl(null);
+    isNavigationAllowedRef.current = false;
+  }, []);
+
+  useEffect(() => {
+    if (!router?.events || !router.isReady) return;
+
+    const handleRouteChangeError = (err: unknown): void => {
+      if ((err as { cancelled?: boolean })?.cancelled) {
+        return;
+      }
+      throw err as Error;
+    };
+
+    const handleRouteChangeStart = (url: string): void => {
+      if (!hasUnsavedChangesRef.current || isNavigationAllowedRef.current) {
+        isNavigationAllowedRef.current = false;
+        return;
+      }
+
+      setPendingUrl(url);
+      setIsBlocked(true);
+
+      const err: Error & { cancelled?: boolean } = new Error(
+        "Navigation blocked due to unsaved changes",
+      );
+      err.cancelled = true;
+      router.events.emit("routeChangeError", err, url, { shallow: false });
+      throw err;
+    };
+
+    router.events.on("routeChangeError", handleRouteChangeError);
+    router.events.on("routeChangeStart", handleRouteChangeStart);
+
+    return (): void => {
+      router.events.off("routeChangeError", handleRouteChangeError);
+      router.events.off("routeChangeStart", handleRouteChangeStart);
+    };
+  }, [router]);
+
+  // Handle browser refresh/close
+  useEffect(() => {
+    const handleBeforeUnload = (e: BeforeUnloadEvent): string | undefined => {
+      if (hasUnsavedChanges) {
+        e.preventDefault();
+        // For cross-browser support of the built-in "Leave site?" dialog, set returnValue in addition to preventDefault.
+        // See MDN: https://developer.mozilla.org/en-US/docs/Web/API/Window/beforeunload_event
```

Comment:
I left this note here for whoever looks at this in the future because TypeScript shows e.returnValue as being deprecated, but apparently it's recommended for legacy cross-browser support